src/main/java/com/streamr/client/StreamrClient.java

```diff
@@ -190,81 +188,48 @@ public WebSocketClient getWebsocket() {
-    }
-
-    private void connect(boolean firstTrial) throws ConnectionTimeoutException {
         if (getState() == ReadyState.OPEN) {
-            log.warn("Trying to connect when already connected to " + options.getWebsocketApiUrl());
             return;
         }
 
-        if (firstTrial) {
-            log.info("Connecting to " + options.getWebsocketApiUrl() + "...");
-            if (websocket == null) {
-                initWebsocket();
+        if (!keepConnected) {
+            synchronized (lock) {
+                keepConnected = true;
+                log.info("Connecting to " + options.getWebsocketApiUrl() + "...");
+                executorService.scheduleAtFixedRate(() -> {
+                            if (keepConnected) {
+                                if (getState() != ReadyState.OPEN) {
+                                    boolean isReconnect = this.websocket != null;
+                                    log.info("Not connected. Attempting to " + (isReconnect ? "reconnect" : "connect"));
+                                    if (isReconnect) {
+                                        this.websocket.closeConnection(0, "");
+                                    }
+                                    initWebsocket();
+                                    this.websocket.connect();
+                                }
+                            } else {
+                                if (getState() != ReadyState.CLOSED) {
+                                    log.info("Closing connection");
+                                    websocket.closeConnection(0, "");
+                                    websocket = null;
+                                    executorService.shutdown();
+                                }
+                            }
+                        },
+                        0,
+                        options.getReconnectRetryInterval(),
+                        TimeUnit.MILLISECONDS
+                );
             }
-            websocket.connect();
-            waitForState(ReadyState.OPEN);
-        } else {
-            log.info("Reconnecting to " + options.getWebsocketApiUrl() + "...");
-            reconnect();
         }
 
-        if (errorWhileConnecting != null) {
-            Exception ex = errorWhileConnecting;
-            errorWhileConnecting = null;
-            throw new RuntimeException(ex);
-        } else if (getState() != ReadyState.OPEN) {
-            log.warn("Failed to connect to " + options.getWebsocketApiUrl() + ". Going to retry in " + options.getReconnectRetryInterval() / 1000 + " seconds.");
-            try {
-                Thread.sleep(options.getReconnectRetryInterval());
-            } catch (InterruptedException e) {
-                e.printStackTrace();
-            }
-            connect(false);
+        waitForState(ReadyState.OPEN);
+        if (getState() != ReadyState.OPEN) {
+            throw new ConnectionTimeoutException(options.getWebsocketApiUrl());
```

Comment:
Should we shutdown the scheduled executor here before throwing?

Reply:
The idea here is that even if the initial connect fails to go thru in the time that `waitForState` waits, we still want to connect whenever it is possible. So calling `connect` in a sense puts the client in the mode "keep connected". For that to happen, the executor must be alive.

The waiting for `ReadyState.OPEN` is thus just merely a convenience feature.

Reply:
I'm thinking of making a second PR which gets rid of the waiting altogether. As we know, a connection may be lost at any point. So I don't know if there is much reason to have special treatment of this initial connection when a re-connection may occur later anyway. 

The key thing to solve imo is that published messages are queued while disconnected. Thus eventual delivery is guaranteed unless the internal queue goes full. We could return a Future or something from the `publish` method if a user really wants to make sure that the message went thru the WebSocket and to catch potential delivery failures.

Reply:
> The key thing to solve imo is that published messages are queued while disconnected.

I agree, this is probably the key thing. `connect() + publish()` is a very common pattern, and currently the behavior has been ok, as `connect()` would block until connected, and subsequent calls to `publish()` would then work because we're connected. If we give up the blocking of `connect()`, then we should have the publish queue.

But for the comment itself: if the client keeps connecting, I'm wondering whether throwing the `ConnectionTimeoutException` to the user makes sense. To me as a user, it would signal than the connection failed (permanently) and everything is halted. But in reality, the reconnect attempts continue.

For now I think we can leave it like this, if we plan to improve it in the future.